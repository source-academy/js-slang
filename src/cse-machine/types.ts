import * as es from 'estree'

import { Environment } from '../types'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
  srcNode: es.Node
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
