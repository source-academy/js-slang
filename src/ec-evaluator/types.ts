import * as es from 'estree'

import { Context } from '..'
import { Environment } from '../types'
import { Agenda, Stash } from './interpreter'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  PUSH_UNDEFINED_IF_NEEDED = 'PushUndefinedIfNeeded',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
  srcNode: es.Node
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
  srcNode: es.Node
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
  srcNode: es.Node
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
  srcNode: es.Node
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

export type CmdEvaluator = (
  command: AgendaItem,
  context: Context,
  agenda: Agenda,
  stash: Stash
) => void

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
