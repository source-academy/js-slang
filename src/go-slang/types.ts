export enum NodeType {
  ExpressionStatement = 'ExpressionStatement',
  Literal = 'Literal',
  UnaryExpression = 'UnaryExpression',
  BinaryExpression = 'BinaryExpression'
}

// NOTE: right now, the only type of source file is an expression statement
export type SourceFile = ExpressionStatement

type Expression = Literal | UnaryExpression | BinaryExpression

export interface Node {
  type: NodeType
}

export interface ExpressionStatement extends Node {
  type: NodeType.ExpressionStatement
  expression: Expression
}

export interface Literal extends Node {
  type: NodeType.Literal
  value: any
}

export type UnaryOperator = '+' | '-'

export interface UnaryExpression extends Node {
  type: NodeType.UnaryExpression
  operator: UnaryOperator
  argument: Expression
}

export type BinaryOperator =
  | '+'
  | '-'
  | '*'
  | '/'
  | '%'
  | '|'
  | '^'
  | '=='
  | '!='
  | '<'
  | '<='
  | '>'
  | '>='

export interface BinaryExpression extends Node {
  type: NodeType.BinaryExpression
  operator: BinaryOperator
  left: Expression
  right: Expression
}

export enum CommandType {
  UnaryOp = 'UnaryOp',
  BinaryOp = 'BinaryOp'
}

export interface Command {
  type: CommandType
}

export interface UnaryOp extends Command {
  type: CommandType.UnaryOp
  operator: UnaryOperator
}

export interface BinaryOp extends Command {
  type: CommandType.BinaryOp
  operator: BinaryOperator
}

export type Instruction = ExpressionStatement | Expression | UnaryOp | BinaryOp
