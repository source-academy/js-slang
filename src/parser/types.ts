import type { Program } from 'estree'

import type { Context, Chapter, Node, SourceError, Variant } from '../types'

export type { Options as AcornOptions } from 'acorn'
export type { ParserOptions as BabelOptions } from '@babel/parser'

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}

export interface Rule<T extends Node> {
  name: string
  disableFromChapter?: Chapter
  disableForVariants?: Variant[]
  checkers: {
    [name: string]: (node: T, ancestors: Node[]) => SourceError[]
  }
}
