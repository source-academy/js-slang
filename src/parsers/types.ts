import { ParserOptions } from '@babel/parser'
import { Options } from 'acorn'
import { Program } from 'estree'

import { Context } from '../types'

export type AcornOptions = Options
export type BabelOptions = ParserOptions

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}
