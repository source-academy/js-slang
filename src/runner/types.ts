import type { Program } from 'estree'
import type { Context, IOptions, Result } from '..'

export type Runner = (program: Program, context: Context, options: IOptions) => Promise<Result>
