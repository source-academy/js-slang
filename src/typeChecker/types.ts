// =======================================
// Types used in type checker for type inference/type error checker for Source Typed variant
// =======================================

import type es from 'estree'
import type { AllowedDeclarations, Node } from '../utils/ast/node'

export type PrimitiveType = 'boolean' | 'null' | 'number' | 'string' | 'undefined'
export type TSAllowedTypes = 'any' | 'void'
export const disallowedTypes = ['bigint', 'never', 'object', 'symbol', 'unknown'] as const

export type TSDisallowedTypes = (typeof disallowedTypes)[number]

/**
 * All types recognised by type parser as basic types
 */
export type TSBasicType = PrimitiveType | TSAllowedTypes | TSDisallowedTypes

// Types for nodes used in type inference
export type NodeWithInferredType<T extends Node> = InferredType & T

export type InferredType = Untypable | Typed | NotYetTyped

export interface TypedFuncDecl {
  functionInferredType?: Type
}
export interface Untypable {
  typability?: 'Untypable'
  inferredType?: Type
}
export interface NotYetTyped {
  typability?: 'NotYetTyped'
  inferredType?: Type
}
export interface Typed {
  typability?: 'Typed'
  inferredType?: Type
}

// Constraints used in type inference
export type Constraint = 'none' | 'addable'

// Types used by both type inferencer and Source Typed
export type Type =
  | Primitive
  | Variable
  | FunctionType
  | List
  | Pair
  | SArray
  | UnionType
  | LiteralType

export interface Primitive {
  kind: 'primitive'
  name: PrimitiveType | TSAllowedTypes
  // Value is needed for Source Typed type error checker due to existence of literal types
  value?: string | number | boolean
}

// In Source Typed, Variable type is used for
// 1. Type parameters
// 2. Type references of generic types with type arguments
export interface Variable {
  kind: 'variable'
  name: string
  constraint: Constraint
  // Used in Source Typed variant to store type arguments of generic types
  typeArgs?: Type[]
}

// cannot name Function, conflicts with TS
export interface FunctionType {
  kind: 'function'
  parameterTypes: Type[]
  returnType: Type
}

export interface List {
  kind: 'list'
  elementType: Type
  // Used in Source Typed variants to check for type mismatches against pairs
  typeAsPair?: Pair
}
export interface Pair {
  kind: 'pair'
  headType: Type
  tailType: Type
}

export interface SArray {
  kind: 'array'
  elementType: Type
}

// Union types and literal types are only used in Source Typed for typechecking
export interface UnionType {
  kind: 'union'
  types: Type[]
}
export interface LiteralType {
  kind: 'literal'
  value: string | number | boolean
}

// In Source Typed, ForAll type is used for generic types
export interface ForAll {
  kind: 'forall'
  polyType: Type
  // Used in Source Typed variant to store type parameters of generic types
  typeParams?: Variable[]
}
export interface PredicateType {
  kind: 'predicate'
  ifTrueType: Type | ForAll
}
export type BindableType = Type | ForAll | PredicateType
export type PredicateTest = {
  node: NodeWithInferredType<es.CallExpression>
  ifTrueType: Type | ForAll
  argVarName: string
}

/**
 * Each element in the TypeEnvironment array represents a different scope
 * (e.g. first element is the global scope, last element is the closest).
 * Within each scope, variable types/declaration kinds, as well as type aliases, are stored.
 */
export type TypeEnvironment = {
  typeMap: Map<string, BindableType>
  declKindMap: Map<string, AllowedDeclarations>
  typeAliasMap: Map<string, Type | ForAll>
}[]
