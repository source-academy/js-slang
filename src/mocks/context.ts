import * as es from 'estree'

import createContext from '../createContext'
import Closure from '../interpreter/closure'
import { Context, Environment, Variant } from '../types'

export function mockContext(chapter = 1, variant: Variant = 'default'): Context {
  return createContext(chapter, variant)
}

export function mockRuntimeContext(): Context {
  const context = createContext()
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ]
  }
  return context
}

export function mockClosure(): Closure {
  return new Closure(
    {
      type: 'FunctionExpression',
      loc: null,
      id: null,
      params: [],
      body: {
        type: 'BlockStatement',
        body: []
      }
    } as es.FunctionExpression,
    {} as Environment,
    {} as Context
  )
}
