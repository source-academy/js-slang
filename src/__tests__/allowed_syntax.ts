import { parseError, runInContext } from '../index'
import { mockContext } from '../mocks/context'
import { Finished } from '../types'

test('Syntaxes are allowed in the chapter they are introduced', () => {
  const code = [
    [
      1,
      `
      `
    ],

    [
      1,
      `
      function name(a, b) {
        const sum = a + b;
        if (sum > 1) {
          return sum;
        } else {
          if (a % 2 === 0) {
            return -1;
          } else if (b % 2 === 0) {
            return 1;
          } else {
            return a > b ? 0 : -2;
          }
        }
      }
      name(1, 2);
      `
    ],

    [
      1,
      `
      (() => true)();
      `
    ],

    [
      1,
      `
      ((x, y) => { return x + y; })(1, 2);
      `
    ],

    [
      1,
      `
      true;
      `
    ],

    [
      1,
      `
      false;
      `
    ],

    [
      1,
      `
      'a string "" \\'\\'';
      `
    ],

    [
      1,
      `
      31.4 + (-3.14e10) * -1 % 2 / 1.5;
      `
    ],

    [
      1,
      `
      !false === (1 !== 2) && 1 < 2 && 1 <= 2 && 2 >= 1 && 2 > 1 || false;
      `
    ],

    [
      1,
      `
      true ? 1 : 2;
      `
    ],

    [
      2,
      `
      null;
      `
    ],

    [
      2,
      `
      pair(1, null);
      `
    ],

    [
      2,
      `
      list(1);
      `
    ],

    [
      3,
      `
      let i = 1;
      while (i < 5) {
        i = i + 1;
      }
      i;
      `
    ],

    [
      3,
      `
      let i = 1;
      for (i = 1; i < 5; i = i + 1) {
      }
      i;
      `
    ],

    [
      3,
      `
      let i = 1;
      for (let j = 0; j < 5; j = j + 1) {
        if (j < 1) {
          continue;
        } else {
          i = i + 1;
          if (j > 2) {
            break;
          } else {
          }
        }
      }
      i;
      `
    ],

    [
      3,
      `
      [];
      `
    ],

    [
      3,
      `
      [1, 2, 3];
      `
    ],

    [
      3,
      `
      [1, 2, 3][1];
      `
    ],

    [
      3,
      `
      let x = [1, 2, 3];
      x[1];
      `
    ],

    [
      3,
      `
      let x = [1, 2, 3];
      x[1] = 4;
      `
    ],

    [
      100,
      `
      ({});
      `
    ],

    [
      100,
      `
      ({a: 1, b: 2});
      `
    ],

    [
      100,
      `
      ({a: 1, b: 2})['a'];
      `
    ],

    [
      100,
      `
      ({a: 1, b: 2}).a;
      `
    ],

    [
      100,
      `
      ({'a': 1, 'b': 2}).a;
      `
    ],

    [
      100,
      `
      ({1: 1, 2: 2})['1'];
      `
    ],

    [
      100,
      `
      const key = 'a';
      ({a: 1, b: 2})[key];
      `
    ],

    [
      100,
      `
      let x = {a: 1, b: 2};
      x.a = 3;
      `
    ],

    [
      100,
      `
      let x = {a: 1, b: 2};
      x['a'] = 3;
      `
    ],

    [
      100,
      `
      let x = {a: 1, b: 2};
      const key = 'a';
      x[key] = 3;
      `
    ]
  ]

  const scheduler = 'preemptive'
  const promises = code.map(c => {
    const chapter = c[0] as number
    const snippet = c[1] as string
    const parseSnippet = `parse(${JSON.stringify(snippet)});`
    const successContext = mockContext(chapter)
    const parseContext = mockContext(Math.max(4, chapter))
    const failureContext = mockContext(chapter - 1)
    return runInContext(snippet, successContext, { scheduler }).then(runResult =>
      runInContext(parseSnippet, parseContext, { scheduler }).then(parseResult =>
        runInContext(snippet, failureContext, { scheduler }).then(failureResult => ({
          chapter,
          snippet,
          parseSnippet,
          successContext,
          parseContext,
          failureContext,
          runResult,
          parseResult,
          failureResult
        }))
      )
    )
  })
  return Promise.all(promises).then(results => {
    results.map(res => {
      const {
        chapter,
        snippet,
        parseSnippet,
        successContext,
        parseContext,
        failureContext,
        runResult,
        parseResult,
        failureResult
      } = res
      const successErrors = parseError(successContext.errors)
      const parseErrors = parseError(parseContext.errors)
      const failureErrors = parseError(failureContext.errors)

      // If you hit an error here, you have changed the snippets but not changed the snapshot
      expect(snippet).toMatchSnapshot()
      expect(parseSnippet).toMatchSnapshot()

      expect(successErrors).toMatchSnapshot()
      expect(parseErrors).toMatchSnapshot()
      expect(runResult.status).toBe('finished')
      expect((runResult as Finished).value).toMatchSnapshot()
      expect(parseResult.status).toBe('finished')
      expect((parseResult as Finished).value).toMatchSnapshot()

      if (chapter > 1) {
        expect(failureErrors).toMatchSnapshot()
        expect(failureResult.status).toBe('error')
      }
    })
  })
})
