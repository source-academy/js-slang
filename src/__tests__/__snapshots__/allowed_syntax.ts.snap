// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Syntaxes are allowed in the chapter they are introduced 1`] = `
"
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 2`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 3`] = `undefined`;

exports[`Syntaxes are allowed in the chapter they are introduced 4`] = `
"
    function name(a, b) {
      const sum = a + b;
      if (sum > 1) {
        return sum;
      } else {
        if (a % 2 === 0) {
          return -1;
        } else if (b % 2 === 0) {
          return 1;
        } else {
          return a > b ? 0 : -2;
        }
      }
    }
    name(1, 2);
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 5`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 6`] = `3`;

exports[`Syntaxes are allowed in the chapter they are introduced 7`] = `
"
    (() => true)();
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 8`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 9`] = `true`;

exports[`Syntaxes are allowed in the chapter they are introduced 10`] = `
"
    ((x, y) => { return x + y; })(1, 2);
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 11`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 12`] = `3`;

exports[`Syntaxes are allowed in the chapter they are introduced 13`] = `
"
    true;
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 14`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 15`] = `true`;

exports[`Syntaxes are allowed in the chapter they are introduced 16`] = `
"
    false;
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 17`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 18`] = `false`;

exports[`Syntaxes are allowed in the chapter they are introduced 19`] = `
"
    'a string \\"\\" \\\\'\\\\'';
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 20`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 21`] = `"a string \\"\\" ''"`;

exports[`Syntaxes are allowed in the chapter they are introduced 22`] = `
"
    31.4 + (-3.14e10) * -1 % 2 / 1.5;
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 23`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 24`] = `31.4`;

exports[`Syntaxes are allowed in the chapter they are introduced 25`] = `
"
    !false === (1 !== 2) && 1 < 2 && 1 <= 2 && 2 >= 1 && 2 > 1 || false;
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 26`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 27`] = `true`;

exports[`Syntaxes are allowed in the chapter they are introduced 28`] = `
"
    null;
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 29`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 30`] = `null`;

exports[`Syntaxes are allowed in the chapter they are introduced 31`] = `
"
    pair(1, null);
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 32`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 33`] = `
Array [
  1,
  null,
]
`;

exports[`Syntaxes are allowed in the chapter they are introduced 34`] = `
"
    list(1);
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 35`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 36`] = `
Array [
  1,
  null,
]
`;

exports[`Syntaxes are allowed in the chapter they are introduced 37`] = `
"
    let i = 1;
    while (i < 5) {
      i = i + 1;
    }
    i;
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 38`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 39`] = `5`;

exports[`Syntaxes are allowed in the chapter they are introduced 40`] = `
"
    let i = 1;
    for (i = 1; i < 5; i = i + 1) {
    }
    i;
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 41`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 42`] = `5`;

exports[`Syntaxes are allowed in the chapter they are introduced 43`] = `
"
    let i = 1;
    for (let j = 0; j < 5; j = j + 1) {
      if (j < 1) {
        continue;
      } else {
        i = i + 1;
        if (j > 2) {
          break;
        } else {
        }
      }
    }
    i;
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 44`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 45`] = `4`;

exports[`Syntaxes are allowed in the chapter they are introduced 46`] = `
"
    [];
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 47`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 48`] = `Array []`;

exports[`Syntaxes are allowed in the chapter they are introduced 49`] = `
"
    [1, 2, 3];
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 50`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 51`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`Syntaxes are allowed in the chapter they are introduced 52`] = `
"
    [1, 2, 3][1];
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 53`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 54`] = `2`;

exports[`Syntaxes are allowed in the chapter they are introduced 55`] = `
"
    ({});
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 56`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 57`] = `Object {}`;

exports[`Syntaxes are allowed in the chapter they are introduced 58`] = `
"
    ({a: 1, b: 2});
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 59`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 60`] = `
Object {
  "a": 1,
  "b": 2,
}
`;

exports[`Syntaxes are allowed in the chapter they are introduced 61`] = `
"
    ({a: 1, b: 2})['a'];
  "
`;

exports[`Syntaxes are allowed in the chapter they are introduced 62`] = `""`;

exports[`Syntaxes are allowed in the chapter they are introduced 63`] = `1`;
